### ISSUE TITLE
Multiple actions dispatch cancel underlying HTTP request

### ISSUE THREAD

**Issue by colthreepv** [ngrx/effects#134]
I am trying to pull down a data structure with relations, let's say I have a `Department` that haves a `Member[]` and I will do one AJAX call for each Member

**effects.ts**
```typescript
@Effect() retrieveDepartment$: Observable<Action> = this.action$
  .ofType(Actions.RETRIEVE_DEPARTMENT)
  .withLatestFrom(this.store)
  .switchMap(([action, store]) => {
    const { members }: { members: number[] } = action.payload;
    const availableMembers = store.members;

    const missingMembers = members.filter(member => availableMembers[member] == null);

    return Observable.from(missingMembers)
      .map(member => ({ type: Actions.FETCH_MEMBER, payload: member }));
  });

@Effect() retrieveMember$: Observable<Action> = this.action$
  .ofType(Actions.FETCH_MEMBER)
  .switchMap(action => this.api.getMember(action.payload))
  .map(member => ({ type: Actions.RETRIEVE_MEMBER, payload: member }));
```

The first effect triggers on receiving the `Department` action, it will create a batch of new actions to retrieve the underlying `Members`

The ugly part happens in the `retrieveMember$` as it spews out many HTTP calls that gets instantaneously cancelled, and only lets the last one go.

I suppose this is not a bug with ngrx/effects, just an ugly use case that might happen, and I wonder how to work around that; what I was expecting was to dispatch many asynchronous calls in a parallel way without caring for the order

**Comment by kylecordes** [comment]
Possibly it might help to revamp your actions - RETRIEVE_MEMBER and FETCH_MEMBER suggest that maybe the actions are saying the wrong thing. Actions generally should be able what happened - the decision to undertake fetching as a result, belongs in the effects, rather than in the actions per se. This might just be a naming thing, of course.

For your specific case, changing switchMap to mergeMap will cause all of those requests to keep going, instead of cancelling each other.

But - you may be better off overall, avoiding the 'cascading effects'. Just use one effect which does the right fetching instead?

**Comment by colthreepv** [comment]
> RETRIEVE_MEMBER and FETCH_MEMBER suggest that maybe the actions are
saying the wrong thing. Actions generally should be able what happened -
the decision to undertake fetching as a result, belongs in the effects,
rather than in the actions per se. This might just be a naming thing, of
course.

Good input, I will look deeper in the examples repository for guidelines

> But - you may be better off overall, avoiding the 'cascading effects'.
Just use one effect which does the right fetching instead?

That was to make the UI a bit responsive and not wait for a (possibly) very
long list of AJAX calls.
I have to admit the first thing that I thought was to use a similar
approach as Promise.all

nice catch, It worked flawlessly, I didn't think it was as easy

**Comment by kylecordes** [comment]
@colthreepv Yes - you could use a single effect, and also keep the UI responsive. It is OK for one incoming action to trigger an Effect, which then emits a series of Actions with data that comes back incrementally. Sometimes `.merge(Observable.of(initialBitOfReponse), theRestOfIt)` or similar is helpful. This emits the initial bit right now, and the rest later as it arrives.

**Comment by guenam** [comment]
I had a similar issue: If I remember well, this is by design. The solution was for me to use concat (imported from 'rxjs/add/operator/concat').

**Comment by josh-sachs** [comment]
@guenam any chance you can recall how you used concat to solve?  This seems problematic when the effect is triggered multiple times... The HTTP call gets cancelled. 

`effect$ = actions$.ofType("ACTION").withLatestFrom().switchMap(() => HTTP).map(() => "ON_ACTION")`

**Comment by nmt1994** [comment]
My use case was I have 2/more components that when init it will load dispatch 2 events which has the same type. Using switchMap would cancel the previous action, so for those kind of "actions", I'd use concatMap / mergeMap
```javascript
effect$ = action$.ofType('ACTION').concatMap( 
    ( action : Action ) => console.log(action);
)
```

**Comment by franciscotln** [comment]
by design, `switchMap` has a built-in cancelation system, so when you have higher-order observables being flattened by `switchMap` it will unsubscribe from the previous one. If you use `mergeMap` (also known as `flatMap`) you won't have this problem.
If the order of the promises being resolved is important to you then you should use `concatMap` but there is a catch here. Let's say you have 3 requests to the API:
req1 => resolves after 10 seconds
req2 => resolves after 1 second
req3 => resolves after 0.5 second
Using `concatMap` you will only get values when all observables (all promises) emit their values. In this case although req2 and req3 are already resolved after 1 second your subscriber won't get values until req1 gets resolved 9 seconds later.
On the other hand if the order is not important to you and you do not want to cancel previous observables after spawning a new one then you should go with `mergeMap` because in this example that I gave above your subscriber would see:
req3 => first value arrived from the API after half second
req2 => second value arrived from the API after 1 second
req1 => third value arrived from the API after 10 seconds

BTW: I'm expecting your call to the API done by `getMember()` to have a `.catch((err) => Observable.of(someAction))` otherwise your observable may die if some of the requests fails to resolve.
